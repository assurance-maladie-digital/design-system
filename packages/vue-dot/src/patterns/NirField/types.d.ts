interface InputFacadeEventTarget extends EventTarget {
	unmaskedValue: string;
}

export interface InputFacadeEvent extends Event {
	target: InputFacadeEventTarget;
}
