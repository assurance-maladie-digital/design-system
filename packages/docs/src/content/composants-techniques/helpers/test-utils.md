---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
