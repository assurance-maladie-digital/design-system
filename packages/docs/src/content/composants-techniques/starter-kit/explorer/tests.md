Le dossier `tests` contient les tests d’interface et les tests unitaires globaux de l’application.

<doc-alert type="warning">
Il n’est pas recommandé de renommer ce dossier.
</doc-alert>
