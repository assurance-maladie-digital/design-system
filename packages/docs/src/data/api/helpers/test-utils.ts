import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
