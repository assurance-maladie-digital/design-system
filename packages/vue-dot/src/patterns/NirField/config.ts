export const config = {
	persistentHint: true,
	validateOnBlur: true,
	icon: {
		color: 'success'
	},
	tooltip: {
		top: true
	}
};
