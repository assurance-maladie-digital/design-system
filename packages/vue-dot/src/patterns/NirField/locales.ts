export const locales = {
	numberLabel: 'Numéro de sécurité sociale',
	numberHint: '13 caractères',
	keyLabel: 'Clé',
	keyHint: '2 chiffres'
};
