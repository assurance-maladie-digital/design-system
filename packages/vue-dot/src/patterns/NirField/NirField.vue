<template>
	<div class="vd-nir-field d-flex align-start mx-n1 mx-sm-n2">
		<VTextField
			ref="number"
			v-facade="numberMask"
			v-bind="textFieldOptions"
			:value="computedNumberValue"
			:label="locales.numberLabel"
			:hint="locales.numberHint"
			:rules="numberRules"
			:success="numberFilled"
			class="vd-number-field flex-grow-0 mx-1 mx-sm-2"
			@keydown="focusKeyField"
			@input.native="setNumberValue"
			@change="emitChangeEvent"
		>
			<template #append>
				<VIcon
					v-if="numberFilled"
					v-bind="options.icon"
				>
					{{ checkIcon }}
				</VIcon>
			</template>
		</VTextField>

		<VTextField
			v-if="!isSingleField"
			ref="key"
			v-facade="keyMask"
			v-bind="textFieldOptions"
			:value="keyValue"
			:label="locales.keyLabel"
			:hint="locales.keyHint"
			:rules="keyRules"
			:success="keyFilled"
			class="vd-key-field flex-grow-0 mx-1 mx-sm-2"
			@keyup.delete="focusNumberField"
			@input.native="setKeyValue"
			@change="emitChangeEvent"
		>
			<template #append>
				<VIcon
					v-if="keyFilled"
					v-bind="options.icon"
				>
					{{ checkIcon }}
				</VIcon>
			</template>
		</VTextField>

		<VTooltip
			v-if="tooltip"
			v-bind="options.tooltip"
		>
			<template #activator="{ on, attrs }">
				<VIcon
					v-bind="attrs"
					class="vd-tooltip-icon mt-4 ml-0 ml-sm-2"
					v-on="on"
				>
					{{ infoIcon }}
				</VIcon>
			</template>

			<slot name="tooltip">
				{{ tooltip }}
			</slot>
		</VTooltip>
	</div>
</template>

<script lang="ts">
	import Vue, { PropType } from 'vue';
	import Component, { mixins } from 'vue-class-component';

	import { config } from './config';
	import { locales } from './locales';
	import { InputFacadeEvent, Refs } from '../../types';

	import { customizable, Options } from '../../mixins/customizable';

	import { required } from '../../rules/required';
	import { exactLength } from '../../rules/exactLength';
	import { ValidationRule } from '../../rules/types';

	import { formatNir } from '../../functions/formatNir';

	import { mdiCheck, mdiInformationOutline } from '@mdi/js';

	import deepMerge from 'deepmerge';

	enum FieldTypesEnum {
		SINGLE = 13,
		DOUBLE = 15
	}

	const NUMBER_LENGTH = 13;
	const KEY_LENGTH = 2;

	const Props = Vue.extend({
		props: {
			value: {
				type: String,
				default: null
			},
			nirLength: {
				type: Number as PropType<FieldTypesEnum>,
				default: FieldTypesEnum.DOUBLE,
				validator(value): boolean {
					return value === FieldTypesEnum.SINGLE || value === FieldTypesEnum.DOUBLE;
				}
			},
			required: {
				type: Boolean,
				default: false
			},
			tooltip: {
				type: String,
				default: undefined
			}
		}
	});

	const MixinsDeclaration = mixins(Props, customizable(config));

	@Component<NirField>({
		inheritAttrs: false,
		model: {
			prop: 'value',
			event: 'change'
		},
		watch: {
			value: {
				handler(value: string | null) {
					if (!value) {
						return;
					}

					if (this.value.length >= FieldTypesEnum.SINGLE) {
						this.numberValue = value;
					}

					if (this.value.length === FieldTypesEnum.DOUBLE) {
						this.numberValue = value.slice(0, -KEY_LENGTH);
						this.keyValue = value.slice(NUMBER_LENGTH, NUMBER_LENGTH + KEY_LENGTH);
					}
				},
				immediate: true
			}
		}
	})
	export default class NirField extends MixinsDeclaration {
		$refs!: Refs<{
			number: HTMLElement;
			key: HTMLElement;
		}>;

		locales = locales;

		checkIcon = mdiCheck;
		infoIcon = mdiInformationOutline;

		numberValue: string | null = null;
		keyValue: string | null = null;

		numberMask = '# ## ## #X ### ###';
		keyMask = '##';

		get textFieldOptions(): Options {
			return deepMerge<Options>(config, this.$attrs);
		}

		get numberFilled(): boolean {
			return this.numberValue?.length === NUMBER_LENGTH;
		}

		get numberRules(): ValidationRule[] {
			const rulesNumber = [];

			if (this.required) {
				rulesNumber.push(required);
			}

			rulesNumber.push(exactLength(NUMBER_LENGTH, true));

			return rulesNumber;
		}

		setNumberValue(event: InputFacadeEvent): void {
			this.numberValue = event.target?.unmaskedValue ?? null;
		}

		get keyFilled(): boolean {
			return this.keyValue?.length === KEY_LENGTH;
		}

		get keyRules(): ValidationRule[] {
			const rulesKey = [];

			if (this.required) {
				rulesKey.push(required);
			}

			rulesKey.push(exactLength(KEY_LENGTH, true));

			return rulesKey;
		}

		setKeyValue(event: InputFacadeEvent): void {
			this.keyValue = event.target?.unmaskedValue ?? null;
		}

		get computedNumberValue(): string | null {
			return this.numberValue ? formatNir(this.numberValue) : null;
		}

		get internalValue(): string | null {
			if (this.isSingleField && this.numberFilled) {
				return this.numberValue;
			}

			if (!this.numberFilled || !this.keyFilled) {
				return null;
			}

			return this.numberValue as string + this.keyValue as string;
		}

		get isSingleField(): boolean {
			return this.nirLength === FieldTypesEnum.SINGLE;
		}

		focusKeyField({ key, altKey, ctrlKey, metaKey, shiftKey }: KeyboardEvent): void {
			const isSingleField = this.isSingleField;
			const notFilled = !this.numberFilled;
			// Don't move focus for combo (eg. Ctrl + A)
			const keyHasModifier = altKey || ctrlKey || metaKey || shiftKey;
			// Don't move focus for other keys (eg. ArrowRight)
			const isNotSingleAlphaNumChar = !/^[a-zA-Z0-9| ]$/.test(key);
			// Don't move focus is content is selected to allow overwrite
			const isContentSelected = document.getSelection()?.toString() === this.computedNumberValue;

			const shouldNotFocus = isSingleField || notFilled || keyHasModifier || isNotSingleAlphaNumChar || isContentSelected;

			if (shouldNotFocus) {
				return;
			}

			this.$refs.key.focus();
		}

		focusNumberField(): void {
			if (this.keyValue?.length !== 0) {
				return;
			}

			this.$refs.number.focus();
		}

		emitChangeEvent(): void {
			if (!this.internalValue) {
				return;
			}

			this.$emit('change', this.internalValue);
		}
	}
</script>

<style lang="scss" scoped>
	.vd-number-field {
		width: 296px;
	}

	.vd-key-field {
		width: 84px;
	}

	// Don't allow resize for these elements
	.vd-nir-field :deep(.v-input__append-inner),
	.vd-key-field,
	.vd-tooltip-icon {
		flex: none;
	}
</style>
