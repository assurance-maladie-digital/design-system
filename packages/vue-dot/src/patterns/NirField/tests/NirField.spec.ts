import Vue from 'vue';
import { Wrapper } from '@vue/test-utils';

import { mountComponent } from '@/tests';

import NirField from '../';

let wrapper: Wrapper<Vue>;

describe('NirField', () => {
	it('renders correctly', () => {
		wrapper = mountComponent(NirField, undefined, true);

		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly with 13 characters', () => {
		wrapper = mountComponent(NirField, {
			propsData: {
				nirLength: 13
			}
		}, true);

		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly with a tooltip', () => {
		wrapper = mountComponent(NirField, {
			propsData: {
				tooltip: 'Information sur le NIR'
			}
		}, true);

		expect(wrapper).toMatchSnapshot();
	});
});
