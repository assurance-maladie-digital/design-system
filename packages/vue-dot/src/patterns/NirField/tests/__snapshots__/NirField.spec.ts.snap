// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NirField renders correctly 1`] = `
<div class="vd-nir-field d-flex align-start mx-n1 mx-sm-n2">
  <div class="v-input vd-number-field flex-grow-0 mx-1 mx-sm-2 theme--light v-text-field">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-text-field__slot"><label for="input-2" class="v-label theme--light" style="left: 0px; position: absolute;">Numéro de sécurité sociale</label><input icon="[object Object]" tooltip="[object Object]" id="input-2" type="text"></div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
            <div class="v-messages__message">13 caractères</div>
          </transition-group-stub>
        </div>
      </div>
    </div>
  </div>
  <div class="v-input vd-key-field flex-grow-0 mx-1 mx-sm-2 theme--light v-text-field">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-text-field__slot"><label for="input-5" class="v-label theme--light" style="left: 0px; position: absolute;">Clé</label><input icon="[object Object]" tooltip="[object Object]" id="input-5" type="text"></div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
            <div class="v-messages__message">2 chiffres</div>
          </transition-group-stub>
        </div>
      </div>
    </div>
  </div>
  <!---->
</div>
`;

exports[`NirField renders correctly with 13 characters 1`] = `
<div class="vd-nir-field d-flex align-start mx-n1 mx-sm-n2">
  <div class="v-input vd-number-field flex-grow-0 mx-1 mx-sm-2 theme--light v-text-field">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-text-field__slot"><label for="input-10" class="v-label theme--light" style="left: 0px; position: absolute;">Numéro de sécurité sociale</label><input icon="[object Object]" tooltip="[object Object]" id="input-10" type="text"></div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
            <div class="v-messages__message">13 caractères</div>
          </transition-group-stub>
        </div>
      </div>
    </div>
  </div>
  <!---->
  <!---->
</div>
`;

exports[`NirField renders correctly with a tooltip 1`] = `
<div class="vd-nir-field d-flex align-start mx-n1 mx-sm-n2">
  <div class="v-input vd-number-field flex-grow-0 mx-1 mx-sm-2 theme--light v-text-field">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-text-field__slot"><label for="input-15" class="v-label theme--light" style="left: 0px; position: absolute;">Numéro de sécurité sociale</label><input icon="[object Object]" tooltip="[object Object]" id="input-15" type="text"></div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
            <div class="v-messages__message">13 caractères</div>
          </transition-group-stub>
        </div>
      </div>
    </div>
  </div>
  <div class="v-input vd-key-field flex-grow-0 mx-1 mx-sm-2 theme--light v-text-field">
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-text-field__slot"><label for="input-18" class="v-label theme--light" style="left: 0px; position: absolute;">Clé</label><input icon="[object Object]" tooltip="[object Object]" id="input-18" type="text"></div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light">
          <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
            <div class="v-messages__message">2 chiffres</div>
          </transition-group-stub>
        </div>
      </div>
    </div>
  </div> <span class="v-tooltip v-tooltip--top"><!----><span aria-hidden="true" aria-haspopup="true" aria-expanded="false" class="v-icon notranslate vd-tooltip-icon mt-4 ml-0 ml-sm-2 theme--light"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" role="img" aria-hidden="true" class="v-icon__svg"><path d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z"></path></svg></span></span>
</div>
`;
